### ISSUE TITLE
`resolve` equivalent in new router

### ISSUE THREAD

**Issue by PascalPrecht** [angular/router#100]
I read the source code and also the existing documentation you wrote for the router. Since I try to also sort of compare the old with the new router in the article, I need to explain the `resolve` $routeProvider config equivalent of the new router.

As far as I understand, instead of having this `resolve` block, we use `canActivate` and `activate` for it. Am I getting this right?

Whereas `canActivate` decides with the routing happens at all, `activate` is in charge of doing all the work that is needed to make needed data available in the controller. As you show in the docs, this could be something like:

```js
MyController.prototype.activate = function() {
  return this.bigFiles = this.$http.downloadBigFiles();
};
```
`this.bigFiles` is what we usually would request in a `resolve` block. Please correct me if I'm wrong.

However, I still wonder how this behaves, if we have asynchronous work todo (actually the example looks asynchronous too). In the old router, `resolve` properties were injected into the controller that is associated to the view.

I understand that we don't need that anymore, as long as we just expose data on `this` (in the example `this.bigFiles`), since controller methods can just access it. But what if we have a promise, that needs to resolve first? Is that functionality also (still) supported in the new router?

Hope my question is clear.

**Comment by PascalPrecht** [comment]
I've updated the issue title, so it can be properly used as FAQ when searching by labels.

**Comment by andrezero** [comment]
Hi 

It's really cool to have these convention based `can` methods invoked in the controllers but it's not an alternative for some use cases where having a `resolve` mechanism defined within the route (and not the controller) is very useful.

For instance, I use resolve for asynchronous authorisation of the requested route+params. These are properties of the router and not the controller and I can assemble two different apps using the same controllers with different authorisation schemes. 

Cheers

**Comment by niemyjski** [comment]
Not to change the subject but I really never liked the use of resolve for anything more than a parameter value that would already be in the route/state params...

So in your case you are resolving big files, and now you are putting logic that knows how to populate that in your router logic... What do you do when big files changes (aka you know it changes from someplace / you get a notification via websocket). Now your only option is to have the get code for big files in two spots or reload the whole view and all children views... Unless, there is an easy way to accomplish reloading your resolved data without destroying the controller and view... I think we should avoid against using resolve as it could promote a bad practice...

@btford, @PascalPrecht  Thoughts??

**Comment by kpgarrod** [comment]
John Papa's style guide (https://github.com/johnpapa/angular-styleguide#style-y081) recommends the use of route resolves in certain circumstances and provides some good reasons.

**Comment by PascalPrecht** [comment]
@niemyjski Thanks for you input. So the `bigFiles` thing I've mentioned is just an example. It could be anything else. It totally makes sense to have a mechanism like `resolve` of `ngRoute` where work needs to be done **before** the view change takes place. E.g. if you have view/component that only makes sense to be loaded **with** some specific data you do want to make sure that it's available before the view change happens.

However, I agree with you that it's a rather not so nice way to populate that kind of logic into router logic (as it is currently the case for `resolve` in `ngRoute`) but that's why I asked the question here. I've asked the same question via mail and @btford didn't say "yes" to `canActivate()`.

So for me it's still a bit unclear.

**Comment by niemyjski** [comment]
Yeah, I'm in the same boat. Even with uirouter, they seem to promote this "bad" practice throughout. It seems like this should be in a style guide.. I'm in agreement with John Papa (a little bit). I like passing in a function / service to the controller to do my lifting. It seems like this should be in a style guide.. I'm in agreement with john papa (a little bit). I like passing in a function / service to the controller to do my lifting. I just find that i'm always using state params more than I probably should..

**Comment by cjolif** [comment]
People that were leveraging `resolve`  to use AMD to load controllers, directives, service on the fly won't be able to leverage canActivate to achieve similar results. So there is definitely some kind of regression here?

**Comment by btford** [comment]
tl;dr – all lifecycle hooks are injectable, but I expect devs will want to use the `activate` hook like `resolve` was used before.

**Comment by PascalPrecht** [comment]
@btford maybe you can show a simple snippet that would work as an equivalent to `resolve`?

"all lifecycle hooks are injectable" doesn't really help me here... what do you mean by that? Lifecycle hooks are `canActivate`, `activate` etc. I don't think you'd inject those somewhere else right?

**Comment by Narretz** [comment]
Hm, I also liked the resolve previous `resolve` system better, because it decoupled the routing from the actual controller.

**Comment by kpgarrod** [comment]
I agree. I think it is particularly useful for reusable components because
you can pass in different data depending on the context.

**Comment by damiengarbarino** [comment]
Any news on the sample to replace the resolve in the AMD use case? In my use case, I'm loading all directives/services/controllers I need and the view controller itself as well in this resolve function. Thanks!

**Comment by heidi-de** [comment]
Resolve was perfect as it allowed me to read necessary configuration from the server via $http before my controllers were instantiated. My app is basically useless without the config being loaded first. 
Using canActivate seemed to be a shoo-in for this due to the ability to cancel the route however I see that the controller is initialised before the promise is resolved.
So basically a resolve like function is still needed or is there a better way to do this pre-application initialisation? Preferably without doing the HTTP request outside of angular and manually bootstrapping!

**Comment by tommck** [comment]
I'm not sure I understand what can't be done with the new life cycle events.  The constructor for the controller will be run, yes, but the UI won't be rendered until activate() is finished, so anything that depended on the data being there would just be inside your activate() method.

Do you have a specific situation where this won't work?  Maybe a [pseudo]code sample?

**Comment by heidi-de** [comment]
I´ve uploaded a quick and dirty project. Maybe I am doing this wrong, but it is similar to what I am trying to achieve in the proper application. 
https://github.com/heidi-de/new-router-test
Ideally I want the initialisation of the component controller to be delayed until the promise resolves, but as you can see when it runs, the component controller is run first, the promise will become resolved, but the jsontest variable is never updated with the ip address. 
Apologies for the narcissistic component naming :-)
In my real app I need the configuration object in order to set the apps language and also lay out the views based on the which roles the user has.

**Comment by tommck** [comment]
your code has a bug in it.. you're setting "this.jsontest" equal to a Promise object.  I'm sending a pull request with a tweak or two

**Comment by heidi-de** [comment]
Merged your changes and fixed the bug with regards to the promise. I will have another look at my codebase tomorrow and see if I can refactor it some other way with this in mind! Thanks for your help Tom, really appreciate it!

**Comment by tommck** [comment]
No problem at all! Glad it helped

**Comment by heidi-de** [comment]
Looks like I can refactor my app to get around this, but it is a ton of work (my app is very large!). The old way was a lot more convenient, but the advantages of the new way far outweigh the convenience of the old way :-)

**Comment by sebald** [comment]
@PascalPrecht I can see uses for both. `active` seems to be correct place to load data from a server. But in some apps you may have to check first if you can load the data at all. E.g. a 3rd party Github app. If the user requests data for a repository Github may answer with a 404. In this case it would make much more sense to me to load the data in the `canActivate` and have the user redirected to a "You have no permission to do that" page if the loading fails.

**Comment by demisx** [comment]
@sebald :+1: We use a very similar approach in our ng 1.x app. Only we called this function `canAccess`. :) Should probably refactor it to `canActivate`.

**Comment by paulwalker** [comment]
Putting aside the merits of a `resolve` semantic defined on the route vs a `canActivate` defined on the controller, am I correct in my assessment of this conversation that a fundamental issue exists with the new router in that the `canActivate` does not allow for returning a promise?

If so, is there any reasonable technique in the router to run an async process before allowing navigation?

**Comment by btford** [comment]
> canActivate does not allow for returning a promise

canActivate should allow returning a promise. If it does not, then there is a bug.

**Comment by paulwalker** [comment]
Thank you. To touch on the design merits quickly, it would be really nice to have the ability to provide a promise to a group of routes, in which the routes are only registered or able to be used if the promise resolves to true.

I hope that makes sense, perhaps there is a better practice for this sort of thing like only registering certain routes upon async callback confirmation?

**Comment by franleplant** [comment]
This is a really interesting issue, and here are some concerns that come from real world scenarios:

- Former `resolve` object was a good way of getting data from the server before kicking off the controller. The canonical example is when the route is `/user/:id` load from the server that user with that id and only then start `userController`. I don't like how the controller, with the new lifecycle hooks needs to be aware of the routeParams. 

- Former `resolve` object was mediocre for handling redirections upon named params restriction, for example, when the user navigates to `sections/:sectionID`, check if the sectionID does match with an existing section and if not then redirect to a set default section.
`ui-router` solves this with [`regexps`](https://github.com/angular-ui/ui-router/wiki/URL-Routing#stateparams-service) to validate named params, they call them Typed Params. How will one handle this scenario with the new router?

- Former `resolve` object was mediocre for handling hierarchical relationships between resolves and views, for example, when user navigates to `/vehicle/:vehicleId/section/:sectionId` you need to get from the server the vehicle with that id and after you get it, you need to get from the server the sections available for that vehicle (of course this is a trivial example, but it has it's uses).
`ui-router` solves this with [nested routes and resolves](https://github.com/angular-ui/ui-router/wiki/Nested-States-%26-Nested-Views). How will one do this with the new router?

**Comment by paulwalker** [comment]
I have not been successful registering routes after an async callback.
```
  function AppController($router, $session) {
    this.session = $session;

    $router.config(routes);
    $session.signedIn(function() {
      $router.config(authRoutes);
    });
  }
```
I attempted to return the promise in the canActivate of the parent AppController like so: 

```
  AppController.prototype.canActivate = function() {
    return this.session.promise;
  };
```
But angular still attempts to unsuccessfully load the child route. Any ideas?

**Comment by dpsthree** [comment]
I think the functionality is still there, it's just hidden in a healthy layer of indirection. Consider the following:

```Javascript
angular.module('app', [])
  .controller('MyController', ['user', 'myService', MyController]);

function MyController(user, myService) {
  //Setup the controller under the assumption that everything is OK
  this.resultsPromise = myService.getSomeData()
  this.resultsPromise.then(function (results) {
    //Assume this will be reached, we will validate using canActivate
    //Manipulate Data as needed here or in activate()
  });
}

MyController.prototype.canActivate = function() {
  return this.resultsPromise.then(function (results) {
    //Do whatever is needed to verify the results are OK to use
    return true
  }); 
};

MyController.prototype.activate = function () {
  //optional, we can do data processing here to keep it out of the controller
  this.resultsPromise.then(function (results) {
    this.data = manipulateResults(results);
  });
}
```

**Comment by paulwalker** [comment]
Right, I just don't want to have to contain that logic in every controller. This is where something like ui-router's $stateChangeStart event is helpful, allowing one to annotate the route with a particular common security context is helpful.

```
$rootScope.$on('$stateChangeStart', 
function(event, toState, toParams, fromState, fromParams){ 
    event.preventDefault(); 
    // transitionTo() promise will be rejected with 
    // a 'transition prevented' error
})
```

**Comment by dpsthree** [comment]
Another thought, it seems like everyone (including myself) has in mind using the routing controller as the logic controller for the remainder of that route. What if we decoupled them. Dedicate a controller as the place to put code concerning routing and a separate controller for the logic of the route. Use the latter with ng-controller or a directive. Move any data needed from the "resolve" mechanism through a shared service. With this in place we could "pollute" the routing controller as much as we want with logic concerning routing, redirecting, auth, etc... and not mix it with the view's controller code. Another benefit of this is that it frees up the logic controller for reuse.

This approach was also useful with UI router when having a reusable controller was desired, but the injectable route resolution parameters got in the way.

@paulwalker this is a nice piece of functionality from ui-router, but is a benefit of its events, not a lack of a resolve mechanism.

**Comment by alfonso-presa** [comment]
Even though I really like the new router controller hooks and I think they are a must have feature, IMHO there should also be some way of replicating the resolve functionality in the new router for the following reasons:

* With canActivate you can't have your dependencies resolved before your controller is instantiated. This complicates the development of the controllers and requires a deep knowledge of the pipeline in order to get things done correctly.
* From a developer perspective and in therms of code readability the method canActivate should be used to determine if the controller can be activated... nothing more and nothing less :-).
* You should be able to set the initial ViewModel contents in a single method and IMHO the best way to do this is the constructor, as it makes the developer confident about nothing being rendered without the dependencies resolved.

My proposal is adding a $resolve property to the controller containing an object a`la old router's/ui-router resolver property. Here you have an example:

```javascript
function SampleController (items) {
    //Items will contain the result of the ItemsService promise once it is resolved.
    this.items = items;
}
SampleController.$resolve = {
    items: 'ItemsService' //ItemsService is a service returning a promise
};

angular.module('myapp').controller('SampleController', SampleController);
```

I've created a PR with this feature https://github.com/angular/router/pull/333

**Comment by franleplant** [comment]
Overall I agree with final objective of #333 

Another missing feature, which is paired with the former resolve mechanism is the `$routeChange{Success,Error,Start}` events, what about them?

**Comment by alfonso-presa** [comment]
@paulwalker @franleplant I guess $routeChange* events functionality or a similar one can be achieved easily in the new router by even decorating the current pipeline methods or by providing a custom pipeline with more events.

You have an example of this in this test: https://github.com/angular/router/blob/master/test/pipeline.es5.spec.js#L19

@dpsthree The problem I see with that approach (this one https://github.com/angular/router/issues/100#issuecomment-101898349) is that you're polluting the view model (the controller itself) with elements that are not meant to be used by the view it self. I guess this kind of stuff is more easy to clean up with ES6/TS but it doesn't feel right in ES5.

I'm considering that in case the resolve functionality is not implemented in the new router, probably the best will be to initialize the view model in the activate method and leave the constructor almost empty. I guess activate can also return a promise to delay the view rendering in case something is required to be in the view model prior to it. What do you think?

**Comment by cjolif** [comment]
Re-reading all of this, I still think I have a regression compare to the old router. I think the `canActivate` solution works when the controller is here. But one of the `resolve` use-case was to load the controller on-demand using for example AMD (to do lazy loading) at the time the navigation occurs. I don't think the new router is covering this use-case because `canActivate` forces the controller to already be available.

**Comment by cjolif** [comment]
As the code seems to have moved to angular/angular should this issue (as well as the other I suppose) move as well?

**Comment by PascalPrecht** [comment]
I'd like to keep those issues here until the new router is in a state that we can test against it and verify.

**Comment by yellow1912** [comment]
Is there any update on this issue? I would really miss the ability to resolve content before showing the view. My use case is that we pre-render lots of things server side, then we send that chunk of template to browser for displaying. 

Without "resolve", every time the user switches to a new route then the screen will show empty view for a second because the request to the server does not return anything yet.

**Comment by ArniDzhan** [comment]
@yellow1912 canActivate can do exactly what resolve used to do, if you return promise it doesn't load the controller until data loaded.
here is an updated project you can use and see some examples from
hope that helps

https://github.com/petebacondarwin/ng1-component-router-demo

**Comment by pavankumarkatakam** [comment]
Im actually use `resolve` for angular module loading asynchronous by `ocLazyLoader`. Is this possible in component router to load components asynchronously with new component router. overall project im using this.

**Comment by pavankumarkatakam** [comment]
Is there any progress on this?

**Comment by vkniazeu** [comment]
@pavankumarkatakam
https://github.com/angular/angular/issues/4015#issuecomment-230121134

**Comment by marcoskubis** [comment]
Any progress about this? I've made some changes in the plunkr example: [plunkr.co](https://plnkr.co/edit/WYbqWgb0BUmMoWzxZNx7?p=preview) (**Click in heroes**)
